pyirf/scripts/make_DL3.py

```diff
@@ -0,0 +1,338 @@
+            # cfg["analysis"]["obs_time"]["value"],
+            # cfg["analysis"]["obs_time"]["unit"],
+            args.obs_time.split(".")[0],
+            args.obs_time.split(".")[-1],
+        ),
+    )  # and create it if necessary
+    if not os.path.exists(outdir):
+        os.makedirs(outdir)
+
+    # =========================================================================
+    #                   READ DL2 DATA AND STORE IT ACCORDING TO GADF
+    # =========================================================================
+
+    # Load FITS data
+    particles = ["gamma", "electron", "proton"]
+    evt_dict = dict()  # Contain DL2 file for each type of particle
+    for particle in particles:
+        if args.debug:
+            print(f"Loading {particle} DL2 data...")
+        infile = os.path.join(indir, template_input_file.format(particle))
+        evt_dict[particle] = read_FITS(config=cfg, infile=infile)
+
+    # =========================================================================
+    #               PRELIMINARY OPERATIONS FOR SPECIFIC PIPELINES
+    # =========================================================================
+
+    # Some pipelines could provide some of the DL2 data in different ways
+    # After this part, DL2 data is supposed to be equivalent, regardless
+    # of the original pipeline.
+
+    # Later we should move this out of here, perhaps under a "utils" module.
+
+    if args.pipeline == "EventDisplay":
+
+        # THETA IS SUPPOSED TO BE IN DL2 DATA ACCORDING TO GADF
+        # EventDisplay provides true and reconstructed directions, so we
+        # calculate it here and we add it to the tables.
+
+        # WARNING:
+        # this is true only for point-source simulations!
+        # in the case of diffuse simulations we must distinguish between
+        # - distance between reconstructed direction and center of the FoV
+        # - distance between reconstructed and true direction
+
+        for particle in particles:
+
+            THETA = angular_separation(
+                evt_dict[particle]["TRUE_AZ"],
+                evt_dict[particle]["TRUE_ALT"],
+                evt_dict[particle]["AZ"],
+                evt_dict[particle]["ALT"],
+            )  # in degrees
+
+            # Add THETA column
+            evt_dict[particle]["THETA"] = THETA
+
+            # print(evt_dict[particle].head(n=5))
+
+    # =========================================================================
+    #                   REST OF THE OPERATIONS (TO BE REFACTORED)
+    # =========================================================================
+
+    # Apply offset cut to proton and electron
+    for particle in ["electron", "proton"]:
+
+        # There seems to be a problem in using pandas from FITS data
+        # ValueError: Big-endian buffer not supported on little-endian compiler
+
+        # I convert to astropy table....
+        # should we use only those?
```

Comment:
They better support units but are generally not as powerful when it comes to statistics and other manipulation.

Also tables support ND rows, pandas cannot handle this.

I would see if we can make it work with tables to have the units.

Reply:
maybe you can open an issue about this part

I would like to avoid modifying further this PR (since does already the job) and split the problems in smaller ones.

Reply:
This PR is huge and adds a bunch of code which is not really in good shape. A better approach would be to split this up into a range of smaller PRs. Otherwise we add a bunch of new code we than have to change during the sprint.

Reply:
I think the changes we discussed here do not prevent to use the script as it is.
It's already usable by people, partly configurable (e.g. energy, significance).
This was the point of the PR.

I understand that it is not in good shape (in the end as you know it comes from an old part of protopipe), but I think it's better to merge this (which is anyway much better than the previous version - actually the previous version wasn't working at all) and let people use it.

We can lay down the next smaller changes (some of which are already listed here) and take it from there.

Reply:
Ok, if this replaces the other scripts, at the very least these have to be deleted. Otherwise we will have three different scripts of which only one is working in the master.

Reply:
yes, the old one from protopipe can be deleted (I kept it actually because I was afraid of having forgotten parts of the code, but I can now delete it).

the LST one is something that @vuillaut added and I think it works because there was already a page for LST usage (I then create one for EventDisplay)

Reply:
What is the difference? Why are two scripts needed? These duplicate much of the code.

This boils back to the same problem. We should **first** think about data structures and implement small functions before working on these large scripts.

Reply:
> What is the difference? Why are two scripts needed? These duplicate much of the code.

I didn't watch into it in detail but from a quick look,
- there is a reader for LSTchain data that we could move to pyirf.io
- plotting code that (in general) in #10 I at least suggest putting under a `pyirf.plot module` (but in this case I see that it comes from ctaplot)
- the main function seems to come from the old script, but I do not know what modifications have been made (I guess not much)

This is something we can solve in a PR for LSTcahin support with @vuillaut 

> This boils back to the same problem. We should **first** think about data structures and implement small functions before working on these large scripts.

Again, I understand your point, but in my (maybe short-sighted opinion) having a material that at least does something (in this case reading EventDisplat DL2 data and plotting the results against the "truth") is something usefull to give people a sense of what is the checkpoint from where to start "sprinting" .

Reply:
I think that with this PR and a very simple integration test we could be in a good position to go crazy and start more focused jobs

Reply:
I have opened issue #28 about the Pandas vs astropy tables discussions, so I leabe those comments for later reference and resolve this

Reply:
Yes, the lstchain related script was a temporary thing to be able to produce the LST IRFs. It is bound to disappear.

Reply:
perhaps we should open an issue to merge the necessary information from that script into this one so (for the time being, minus future refactorings) we can also use LSTchain data and not only EventDisplay